bindings/python/pymongocrypt/state_machine.py

```diff
@@ -0,0 +1,139 @@
+# limitations under the License.
+
+from abc import abstractmethod
+
+from pymongocrypt.binding import lib
+from pymongocrypt.compat import ABC
+from pymongocrypt.errors import MongoCryptError
+
+
+class MongoCryptCallback(ABC):
+    """Callback ABC to perform I/O on behalf of libbmongocrypt."""
+
+    @abstractmethod
+    def kms_request(self, kms_context):
+        """Complete a KMS request.
+
+        :Parameters:
+          - `kms_context`: A :class:`MongoCryptKmsContext`.
+
+        :Returns:
+          None
+        """
+        pass
+
+    @abstractmethod
+    def collection_info(self, database, filter):
+        """Get the collection info for a namespace.
+
+        The returned collection info is passed to libmongocrypt which reads
+        the JSON schema.
+
+        :Parameters:
+          - `database`: The database on which to run listCollections.
+          - `filter`: The filter to pass to listCollections.
+
+        :Returns:
+          The first document from the listCollections command response as BSON.
+        """
+        pass
+
+    @abstractmethod
+    def mark_command(self, database, cmd):
+        """Mark a command for encryption.
+
+        :Parameters:
+          - `database`: The database on which to run this command.
+          - `cmd`: The BSON command to run.
+
+        :Returns:
+          The marked command response from mongocryptd.
+        """
+        pass
+
+    @abstractmethod
+    def fetch_keys(self, filter):
+        """Yields one or more keys from the key vault.
+
+        :Parameters:
+          - `filter`: The filter to pass to find.
+
+        :Returns:
+          A generator which yields the requested keys from the key vault.
+        """
+        pass
+
+    @abstractmethod
+    def insert_data_key(self, data_key):
+        """Insert a data key into the key vault.
+
+        :Parameters:
+          - `data_key`: The data key document to insert.
+
+        :Returns:
+          The _id of the inserted data key document.
+        """
+        pass
+
+    @abstractmethod
+    def close(self):
+        """Release resources."""
+        pass
+
+
+def run_state_machine(ctx, callback):
+    """Run the libmongocrypt state machine until completion.
+
+    :Parameters:
+      - `ctx`: A :class:`MongoCryptContext`.
+      - `callback`: A :class:`MongoCryptCallback`.
+
+    :Returns:
+      The completed libmongocrypt operation.
+    """
+    while True:
+        state = ctx.state
+        # Check for terminal states first.
+        if state == lib.MONGOCRYPT_CTX_ERROR:
+            ctx._raise_from_status()
+        elif state == lib.MONGOCRYPT_CTX_READY:
+            return ctx.finish()
+        elif state == lib.MONGOCRYPT_CTX_DONE:
+            return None
+
+        if state == lib.MONGOCRYPT_CTX_NEED_MONGO_COLLINFO:
+            list_colls_filter = ctx.mongo_operation()
+            coll_info = callback.collection_info(
+                ctx.database, list_colls_filter)
+            if coll_info:
```

Comment:
I assume coll_info can be None? Should there be an else case?

Reply:
Yeah it can be None when the listCollections finds no matching collection. Skipping the `add_mongo_operation_result` and jumping straight to `complete_mongo_operation` is what we're supposed to do: https://jira.mongodb.org/browse/CDRIVER-3261